lib/Server.js

```diff
@@ -39,24 +39,34 @@ if (!process.env.WEBPACK_SERVE) {
 }
 
 class Server {
-  constructor(options = {}, compiler) {
+  // TODO: change this after legacy dev server usage is deprecated and serve package is updated
+  constructor(options = {}, compiler = {}) {
+    this.name = 'webpack-dev-server';
+    this.options = options.hooks ? compiler : options;
+
     // TODO: remove this after plugin support is published
     if (options.hooks) {
       [options, compiler] = [compiler, options];
     }
 
     validate(schema, options, 'webpack Dev Server');
 
-    this.compiler = compiler;
-    this.options = options;
-    this.logger = this.compiler.getInfrastructureLogger('webpack-dev-server');
     this.sockets = [];
     this.staticWatchers = [];
     // Keep track of websocket proxies for external websocket upgrade.
     this.websocketProxies = [];
     // this value of ws can be overwritten for tests
     this.wsHeartbeatInterval = 30000;
 
+    if (compiler) {
+      this.apply(options);
+    }
+  }
+
+  apply(compiler) {
+    this.compiler = compiler;
+    this.logger = this.compiler.getInfrastructureLogger(this.name);
+
     normalizeOptions(this.compiler, this.options);
     this.applyDevServerPlugin();
```

Comment:
Ideally I think we need inline this helper here, let's do it after https://github.com/webpack/webpack-dev-server/pull/3260/ and some refactoring

Anyway good job and thanks

Reply:
Yeah removing `DevServerPlugin` and perform ops directly on child compilers is next step to do in my list.